source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc

```diff
@@ -107,114 +118,187 @@ class TranscoderImpl : public Transcoder {
+}
 
-  disabled_ = proto_config.services().empty();
-  if (disabled_) {
-    return;
+void AsyncReflectionFetcher::startReflectionRpcs() {
+  envoy::config::core::v3::GrpcService config;
+  config.mutable_envoy_grpc()->set_cluster_name(cluster_name_);
+
+  for (std::string service: services_) {
+    std::unique_ptr<Envoy::Grpc::RawAsyncClient> raw_async_client =
+      std::make_unique<Envoy::Grpc::AsyncClientImpl>(cluster_manager_, config, time_source_);
+    Envoy::Grpc::AsyncClient<ServerReflectionRequest, ServerReflectionResponse>* async_client =
+      new Envoy::Grpc::AsyncClient<ServerReflectionRequest, ServerReflectionResponse>(std::move(raw_async_client));
+    async_clients_.emplace_back(async_client);
+
+
+    Http::AsyncClient::StreamOptions options;
+    // TODO(hqin): Make this timeout part of the configuration
+    options.setTimeout(std::chrono::milliseconds(5000));
+    const google::protobuf::MethodDescriptor* server_reflection_method =
+        google::protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+            "grpc.reflection.v1alpha.ServerReflection.ServerReflectionInfo");
+    if (server_reflection_method == nullptr) {
+      throw EnvoyException("Failed to find MethodDescriptor for ServerReflectionInfo");
+    }
+    Envoy::Grpc::AsyncStream<ServerReflectionRequest> async_stream =
+        async_client->start(*server_reflection_method, *this, options);
+
+    ServerReflectionRequest request;
+    request.set_file_containing_symbol(service);
+    async_stream.sendMessage(request, true);
   }
+}
 
-  FileDescriptorSet descriptor_set;
+void AsyncReflectionFetcher::onReceiveMessage(std::unique_ptr<ServerReflectionResponse>&& message) {
+  std::lock_guard<std::mutex> guard_(receive_message_mutex_);
+  const std::string& service_requested = message->original_request().file_containing_symbol();
 
-  switch (proto_config.descriptor_set_case()) {
-  case envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder::
-      DescriptorSetCase::kProtoDescriptor:
-    if (!descriptor_set.ParseFromString(
-            api.fileSystem().fileReadToEnd(proto_config.proto_descriptor()))) {
-      throw EnvoyException("transcoding_filter: Unable to parse proto descriptor");
-    }
-    break;
-  case envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder::
-      DescriptorSetCase::kProtoDescriptorBin:
-    if (!descriptor_set.ParseFromString(proto_config.proto_descriptor_bin())) {
-      throw EnvoyException("transcoding_filter: Unable to parse proto descriptor");
+  // This was not a requested service, so assume this is a duplicate response and ignore it.
+  if (services_.find(service_requested) == services_.end()) {
+    return;
+  }
+
+  for (const std::string& file_descriptor_bytes:
+      message->file_descriptor_response().file_descriptor_proto()) {
+    if (!file_descriptor_set_.add_file()->ParseFromString(file_descriptor_bytes)) {
+      throw EnvoyException(
```

Comment:
Here, we could not throw exception,  when this function is called, it may not have exception capture.  We just have to handle failure nicely.  maybe log an error

Reply:
I was actually wondering about correct semantics here. My intent with throwing the exception was to try to crash Envoy (and thereby let the user know something is wrong). That's at least consistent with the original behavior of the filter, as far as I can tell.

Do we want to try to crash Envoy, or do we prefer to have the plugin be quietly disabled?

Reply:
we could not crash envoy. It is ok to crash envoy if the config is from boostrap_config,  but sometime the config is from xDS, from remote server,  which may pass in a bad config,  we should just not apply that config, but not to crash the envoy. 

In this case, we just never talk init_target as ready.  so envoy will not use this config.